<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1">
<meta name="generator" content="pdoc3 0.11.6">
<title>glam API documentation</title>
<meta name="description" content="A small package for generating
[PGFinder](https://github.com/Mesnage-Org/pgfinder)-compatible glycopeptide
databases …">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/13.0.0/sanitize.min.css" integrity="sha512-y1dtMcuvtTMJc1yPgEqF0ZjQbhnc/bFhyvIyVNb9Zk5mIGtqVaAB1Ttl28su8AvFMOY0EwRbAe+HCLqj6W7/KA==" crossorigin>
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/13.0.0/typography.min.css" integrity="sha512-Y1DYSb995BAfxobCkKepB1BqJJTPrOp3zPL74AWFugHHmmdcvO+C48WLrUOlhGMc0QG7AE3f7gmvvcrmX2fDoA==" crossorigin>
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.9.0/styles/default.min.css" crossorigin>
<style>:root{--highlight-color:#fe9}.flex{display:flex !important}body{line-height:1.5em}#content{padding:20px}#sidebar{padding:1.5em;overflow:hidden}#sidebar > *:last-child{margin-bottom:2cm}.http-server-breadcrumbs{font-size:130%;margin:0 0 15px 0}#footer{font-size:.75em;padding:5px 30px;border-top:1px solid #ddd;text-align:right}#footer p{margin:0 0 0 1em;display:inline-block}#footer p:last-child{margin-right:30px}h1,h2,h3,h4,h5{font-weight:300}h1{font-size:2.5em;line-height:1.1em}h2{font-size:1.75em;margin:2em 0 .50em 0}h3{font-size:1.4em;margin:1.6em 0 .7em 0}h4{margin:0;font-size:105%}h1:target,h2:target,h3:target,h4:target,h5:target,h6:target{background:var(--highlight-color);padding:.2em 0}a{color:#058;text-decoration:none;transition:color .2s ease-in-out}a:visited{color:#503}a:hover{color:#b62}.title code{font-weight:bold}h2[id^="header-"]{margin-top:2em}.ident{color:#900;font-weight:bold}pre code{font-size:.8em;line-height:1.4em;padding:1em;display:block}code{background:#f3f3f3;font-family:"DejaVu Sans Mono",monospace;padding:1px 4px;overflow-wrap:break-word}h1 code{background:transparent}pre{border-top:1px solid #ccc;border-bottom:1px solid #ccc;margin:1em 0}#http-server-module-list{display:flex;flex-flow:column}#http-server-module-list div{display:flex}#http-server-module-list dt{min-width:10%}#http-server-module-list p{margin-top:0}.toc ul,#index{list-style-type:none;margin:0;padding:0}#index code{background:transparent}#index h3{border-bottom:1px solid #ddd}#index ul{padding:0}#index h4{margin-top:.6em;font-weight:bold}@media (min-width:200ex){#index .two-column{column-count:2}}@media (min-width:300ex){#index .two-column{column-count:3}}dl{margin-bottom:2em}dl dl:last-child{margin-bottom:4em}dd{margin:0 0 1em 3em}#header-classes + dl > dd{margin-bottom:3em}dd dd{margin-left:2em}dd p{margin:10px 0}.name{background:#eee;font-size:.85em;padding:5px 10px;display:inline-block;min-width:40%}.name:hover{background:#e0e0e0}dt:target .name{background:var(--highlight-color)}.name > span:first-child{white-space:nowrap}.name.class > span:nth-child(2){margin-left:.4em}.inherited{color:#999;border-left:5px solid #eee;padding-left:1em}.inheritance em{font-style:normal;font-weight:bold}.desc h2{font-weight:400;font-size:1.25em}.desc h3{font-size:1em}.desc dt code{background:inherit}.source > summary,.git-link-div{color:#666;text-align:right;font-weight:400;font-size:.8em;text-transform:uppercase}.source summary > *{white-space:nowrap;cursor:pointer}.git-link{color:inherit;margin-left:1em}.source pre{max-height:500px;overflow:auto;margin:0}.source pre code{font-size:12px;overflow:visible;min-width:max-content}.hlist{list-style:none}.hlist li{display:inline}.hlist li:after{content:',\2002'}.hlist li:last-child:after{content:none}.hlist .hlist{display:inline;padding-left:1em}img{max-width:100%}td{padding:0 .5em}.admonition{padding:.1em 1em;margin:1em 0}.admonition-title{font-weight:bold}.admonition.note,.admonition.info,.admonition.important{background:#aef}.admonition.todo,.admonition.versionadded,.admonition.tip,.admonition.hint{background:#dfd}.admonition.warning,.admonition.versionchanged,.admonition.deprecated{background:#fd4}.admonition.error,.admonition.danger,.admonition.caution{background:lightpink}</style>
<style media="screen and (min-width: 700px)">@media screen and (min-width:700px){#sidebar{width:30%;height:100vh;overflow:auto;position:sticky;top:0}#content{width:70%;max-width:100ch;padding:3em 4em;border-left:1px solid #ddd}pre code{font-size:1em}.name{font-size:1em}main{display:flex;flex-direction:row-reverse;justify-content:flex-end}.toc ul ul,#index ul ul{padding-left:1em}.toc > ul > li{margin-top:.5em}}</style>
<style media="print">@media print{#sidebar h1{page-break-before:always}.source{display:none}}@media print{*{background:transparent !important;color:#000 !important;box-shadow:none !important;text-shadow:none !important}a[href]:after{content:" (" attr(href) ")";font-size:90%}a[href][title]:after{content:none}abbr[title]:after{content:" (" attr(title) ")"}.ir a:after,a[href^="javascript:"]:after,a[href^="#"]:after{content:""}pre,blockquote{border:1px solid #999;page-break-inside:avoid}thead{display:table-header-group}tr,img{page-break-inside:avoid}img{max-width:100% !important}@page{margin:0.5cm}p,h2,h3{orphans:3;widows:3}h1,h2,h3,h4,h5,h6{page-break-after:avoid}}</style>
<script defer src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.9.0/highlight.min.js" integrity="sha512-D9gUyxqja7hBtkWpPWGt9wfbfaMGVt9gnyCvYa+jojwwPHLCzUm5i8rpk7vD7wNee9bA35eYIjobYPaQuKS1MQ==" crossorigin></script>
<script>window.addEventListener('DOMContentLoaded', () => {
hljs.configure({languages: ['bash', 'css', 'diff', 'graphql', 'ini', 'javascript', 'json', 'plaintext', 'python', 'python-repl', 'rust', 'shell', 'sql', 'typescript', 'xml', 'yaml']});
hljs.highlightAll();
/* Collapse source docstrings */
setTimeout(() => {
[...document.querySelectorAll('.hljs.language-python > .hljs-string')]
.filter(el => el.innerHTML.length > 200 && ['"""', "'''"].includes(el.innerHTML.substring(0, 3)))
.forEach(el => {
let d = document.createElement('details');
d.classList.add('hljs-string');
d.innerHTML = '<summary>"""</summary>' + el.innerHTML.substring(3);
el.replaceWith(d);
});
}, 100);
})</script>
</head>
<body>
<main>
<article id="content">
<header>
<h1 class="title">Package <code>glam</code></h1>
</header>
<section id="section-intro">
<p>A small package for generating
<a href="https://github.com/Mesnage-Org/pgfinder">PGFinder</a>-compatible glycopeptide
databases.</p>
<p>Just supply a FASTA file containing the protein(s) of interest,
specify the digestion parameters, a glycosylation motif, and a list of potential
glycan compositions / structures.</p>
<h1 id="include-a-short-code-example-here">INCLUDE A SHORT CODE EXAMPLE HERE!</h1>
<p>We can include a markdown file that shows how to use the web interface /
whatever other documentation seems useful!</p>
<pre><code class="language-py">print(&quot;Test?&quot;)
</code></pre>
</section>
<section>
</section>
<section>
<h2 class="section-title" id="header-variables">Global variables</h2>
<dl>
<dt id="glam.DIGESTIONS"><code class="name">var <span class="ident">DIGESTIONS</span> : dict[str, str]</code></dt>
<dd>
<div class="desc"><p>A dictionary mapping common protein digestion treatments to the regular expressions
that describe their cleavage sites.</p></div>
</dd>
<dt id="glam.GLYCOSYLATION_MOTIFS"><code class="name">var <span class="ident">GLYCOSYLATION_MOTIFS</span> : dict[str, str]</code></dt>
<dd>
<div class="desc"><p>A dictionary mapping common glycosylation types to regular expressions that
describe the sequence motifs they target.</p></div>
</dd>
<dt id="glam.MODIFICATIONS"><code class="name">var <span class="ident">MODIFICATIONS</span> : dict[str, glam._lib.Modification]</code></dt>
<dd>
<div class="desc"><p>A dictionary mapping common peptide modifications to a tuple that describes
their abbreviation, the residues they can modify, and their monoisotopic mass
delta.</p></div>
</dd>
</dl>
</section>
<section>
<h2 class="section-title" id="header-functions">Functions</h2>
<dl>
<dt id="glam.generate_glycopeptides"><code class="name flex">
<span>def <span class="ident">generate_glycopeptides</span></span>(<span>fasta: str,<br>digestion: Regex = '',<br>missed_cleavages: int = 0,<br>min_length: int | None = None,<br>max_length: int | None = None,<br>semi_enzymatic: bool = False,<br>glycans: str | None = None,<br>motif: Regex = '',<br>max_glycans: int | None = None,<br>all_peptides: bool = False,<br>modifications: Iterable[glam._lib.Modification] = [],<br>max_modifications: int | None = None,<br>**kwargs) ‑> list[tuple[str, str]]</span>
</code></dt>
<dd>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def generate_glycopeptides(
    fasta: str,
    digestion: Regex = &#34;&#34;,
    missed_cleavages: int = 0,
    min_length: int | None = None,
    max_length: int | None = None,
    semi_enzymatic: bool = False,
    glycans: str | None = None,
    motif: Regex = &#34;&#34;,
    max_glycans: int | None = None,
    all_peptides: bool = False,
    modifications: Iterable[Modification] = [],
    max_modifications: int | None = None,
    **kwargs,
) -&gt; list[tuple[str, str]]:
    &#34;&#34;&#34;Generates glycopeptides from an input FASTA and CSV file of glycans.

    Parameters
    ----------
    fasta : str,
        FASTA text describing the protein sequence(s) to be digested. Note that this
        function ***does not*** read from the filesystem — you&#39;ll need to first load
        your FASTA file into a `str` using something like `pathlib.Path.read_text`.
    enzyme : str or Pattern[str],
        The name of the enyzme used to digest the protein into peptides. A number of
        enzymes are built-in (see `pyteomics.parser.expasy_rules` and
        `pyteomics.parser.psims_rules`), but a custom (optionally pre-compiled)
        regex describing the enzyme&#39;s cleavage site can also be supplied. For more
        information, see `pyteomics`&#39;s documentation.
    motif : str or Pattern[str],
        A regex describing the sequence motif for glycosylation. Only peptides
        containing this motif will be used to generate the final list of glycopeptides.
        See the `glam.GLYCOSYLATION_MOTIFS` dictionary for common motifs.
    glycans : str,

    missed_cleavages : int, default: 0,
        The maximum number of missed cleavages to allow during digestion
    min_length : int or None, default: None,
        The minimum length peptide to include in the digest output
    max_length : int or None, default: None,
        The maximum length peptide to include in the digest output
    semi : bool, default: False,
    &#34;&#34;&#34;

    proteins = pyteomics.fasta.read(StringIO(fasta), use_index=False)
    loaded_glycans = load_glycans(glycans)

    def generate(protein: Protein) -&gt; tuple[str, str]:
        filename = f&#34;{protein.description}.csv&#34;
        seq = protein.sequence

        peptides = digest_protein(
            seq, digestion, missed_cleavages, min_length, max_length, semi_enzymatic
        )
        # NOTE: Doing this first is important so that `find_glycosylation_sites` is
        # aware of modifications. Otherwise it will be possible to have N-residues that
        # are both modified and glycosylated (which isn&#39;t biologically possible)!
        modified_peptides = modify_peptides(peptides, modifications, max_modifications)
        massive_peptides = peptide_masses(modified_peptides, modifications)
        computed_peptides = find_glycosylation_sites(massive_peptides, motif)
        glycopeptides = build_glycopeptides(
            computed_peptides, loaded_glycans, max_glycans, all_peptides
        )

        csv = convert_to_csv(glycopeptides)
        return (filename, csv)

    return [generate(protein) for protein in proteins]</code></pre>
</details>
<div class="desc"><p>Generates glycopeptides from an input FASTA and CSV file of glycans.</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>fasta</code></strong> :&ensp;<code>str,</code></dt>
<dd>FASTA text describing the protein sequence(s) to be digested. Note that this
function <strong><em>does not</em></strong> read from the filesystem — you'll need to first load
your FASTA file into a <code>str</code> using something like <code>pathlib.Path.read_text</code>.</dd>
<dt><strong><code>enzyme</code></strong> :&ensp;<code>str</code> or <code>Pattern[str],</code></dt>
<dd>The name of the enyzme used to digest the protein into peptides. A number of
enzymes are built-in (see <code>pyteomics.parser.expasy_rules</code> and
<code>pyteomics.parser.psims_rules</code>), but a custom (optionally pre-compiled)
regex describing the enzyme's cleavage site can also be supplied. For more
information, see <code>pyteomics</code>'s documentation.</dd>
<dt><strong><code>motif</code></strong> :&ensp;<code>str</code> or <code>Pattern[str],</code></dt>
<dd>A regex describing the sequence motif for glycosylation. Only peptides
containing this motif will be used to generate the final list of glycopeptides.
See the <code><a title="glam.GLYCOSYLATION_MOTIFS" href="#glam.GLYCOSYLATION_MOTIFS">GLYCOSYLATION_MOTIFS</a></code> dictionary for common motifs.</dd>
<dt><strong><code>glycans</code></strong> :&ensp;<code>str,</code></dt>
<dd>&nbsp;</dd>
<dt><strong><code>missed_cleavages</code></strong> :&ensp;<code>int</code>, default<code>: 0,</code></dt>
<dd>The maximum number of missed cleavages to allow during digestion</dd>
<dt><strong><code>min_length</code></strong> :&ensp;<code>int</code> or <code>None</code>, default<code>: None,</code></dt>
<dd>The minimum length peptide to include in the digest output</dd>
<dt><strong><code>max_length</code></strong> :&ensp;<code>int</code> or <code>None</code>, default<code>: None,</code></dt>
<dd>The maximum length peptide to include in the digest output</dd>
<dt><strong><code>semi</code></strong> :&ensp;<code>bool</code>, default<code>: False,</code></dt>
<dd>&nbsp;</dd>
</dl></div>
</dd>
</dl>
</section>
<section>
</section>
</article>
<nav id="sidebar">
<div class="toc">
<ul>
<li><a href="#include-a-short-code-example-here">INCLUDE A SHORT CODE EXAMPLE HERE!</a></li>
</ul>
</div>
<ul id="index">
<li><h3><a href="#header-variables">Global variables</a></h3>
<ul class="">
<li><code><a title="glam.DIGESTIONS" href="#glam.DIGESTIONS">DIGESTIONS</a></code></li>
<li><code><a title="glam.GLYCOSYLATION_MOTIFS" href="#glam.GLYCOSYLATION_MOTIFS">GLYCOSYLATION_MOTIFS</a></code></li>
<li><code><a title="glam.MODIFICATIONS" href="#glam.MODIFICATIONS">MODIFICATIONS</a></code></li>
</ul>
</li>
<li><h3><a href="#header-functions">Functions</a></h3>
<ul class="">
<li><code><a title="glam.generate_glycopeptides" href="#glam.generate_glycopeptides">generate_glycopeptides</a></code></li>
</ul>
</li>
</ul>
</nav>
</main>
<footer id="footer">
<p>Generated by <a href="https://pdoc3.github.io/pdoc" title="pdoc: Python API documentation generator"><cite>pdoc</cite> 0.11.6</a>.</p>
</footer>
</body>
</html>
